hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanP.java

```diff
@@ -0,0 +1,261 @@
+        NodeEngine nodeEngine = ((HazelcastInstanceImpl) hazelcastInstance).node.nodeEngine;
+
+        traverser = new IMapTraverser(
+                nodeEngine,
+                mapScanMetadata,
+                this.partitions
+        );
+        traverser.init(SimpleExpressionEvalContext.from(context));
+    }
+
+    @Override
+    public boolean complete() {
+        Object[] items;
+        if (pendingItems != null) {
+            items = pendingItems;
+            pendingItems = null;
+        } else {
+            items = traverser.next();
+        }
+        for (; items != null; items = traverser.next()) {
+            if (items.length == 0) {
+                return false;
+            }
+            if (!tryEmit(defaultOrdinal, items)) {
+                pendingItems = items;
+                return false;
+            }
+        }
+        return true;
+    }
+
+
+
+    static class IMapTraverser implements Traverser<Object[]> {
+        private final List<Expression<?>> projections;
+        private final Expression<Boolean> filter;
+        private KeyValueIterator iteratorExec;
+
+        private final MapContainer mapContainer;
+        private final JetMapScanMetadata node;
+        private final Set<Integer> partitions;
+        private SimpleExpressionEvalContext ctx;
+        private MapScanRow row;
+
+        IMapTraverser(
+                @Nonnull final NodeEngine nodeEngine,
+                @Nonnull final JetMapScanMetadata node,
+                @Nonnull final List<Integer> partitions
+
+        ) {
+            this.node = node;
+            MapService mapService = nodeEngine.getService(MapService.SERVICE_NAME);
+            this.mapContainer = mapService.getMapServiceContext().getMapContainer(node.getMapName());
+            this.partitions = new PartitionIdSet(partitions.size(), partitions);
+            this.projections = node.getProjects();
+            this.filter = node.getFilter();
+        }
+
+        @Override
+        public Object[] next() {
+            if (iteratorExec.tryAdvance()) {
+                return prepareRow(
+                        iteratorExec.getKey(),
+                        iteratorExec.getKeyData(),
+                        iteratorExec.getValue(),
+                        iteratorExec.getValueData()
+                );
+            }
+            return null;
+        }
+
+        public void init(SimpleExpressionEvalContext ctx) {
+            this.row = MapScanRow.create(
+                    node.getKeyDescriptor(),
+                    node.getValueDescriptor(),
+                    node.getFieldPaths(),
+                    node.getFieldTypes(),
+                    mapContainer.getExtractors(),
+                    ctx.getSerializationService()
+            );
+            this.ctx = ctx;
+            this.iteratorExec = new MapScanExecIterator(mapContainer, partitions.iterator(), ctx.getSerializationService());
```

Comment:
Seems like `IMapTraverser` is mostly a wrapper around `MapScanExecIterator`. Why not adapt code from there and move logic here? We would get rid of one level of abstraction (which is not very useful here) and a class from IMDG engine (which will be eventually removed anyway).

Reply:
#18693